Hotfixrelease 0.13.2

elbenfreund opened the pull request:
This hotfix fixes several issues that have been introduced with the previous release:

- Fix a bug that would not allways return all partial overlaps for actManager._get_all. This in effect fixes broken "timeframe availability" tests. [LIB-253]
- Remove obsolete references to backports.ConfigParser. [LIB-255]
- Properly deal with raw fact that do/do-not include seconds. [LIB-254]
- explicitly pin the version of ``factory-boy`` to be used to avoid side effects when upstream changes and gets implicitly pulled by ``pytest-factoryboy``.

elbenfreund committed:
[LIB-254] time handling: Properly work with seconds

Previous versions of ``hamster-lib`` tried to be overly accomodating by
omitting seconds when parsing/displaying a times. This caused various
bugs and inconsistencies and is fixed by this commit.
While future versions may provide additional liberalizations our
solution now is the most accurate and strict one. It will be easier to
relax things on top of this than the other way round.

Please note that this also means that 'raw fact' parsing now requires
times to be specified as ``%H:%M:%S``

elbenfreund committed:
Remove remaining 'backports.ConfigParser' references

Closes: LIB-255

elbenfreund committed:
Remove remaining 'backports.ConfigParser' references

Closes: LIB-255

elbenfreund committed:
Add testcase to mimic bug-case

elbenfreund committed:
sqlalchemy: add method to check timeframe availability

This commit adds a new private method to the fact manager that handles
checking wether or not a the timeframe for a given fact is available.
This private method is also able to account for the passed fact being
the only fact taking up its own timeframe (relevant for update calls).
This allows for cleaner code in ``FactManager._add`` and
``FactManager._update``.

Future commits should rearange tests accordingly.

Ref: LIB-253

elbenfreund committed:
Clarify storage API docs

elbenfreund committed:
Make dedicated check if timeframe is free method run its own
query

elbenfreund committed:
Adjust tests

elbenfreund committed:
Fix docstrings

elbenfreund committed:
Remove obsolete import, insert linebreak

elbenfreund committed:
Fix spelling as per review

elbenfreund committed:
Refactor timeframe availability check as per suggestion

elbenfreund committed:
Merge branch 'bugfix/253' into release/0.13.2o

FactManager._get_all of the SQLAlchemy backend did not always return all
partial overlaps even if partial=True. As a consequence it was possible
to create new facts that overlapped with existing ones.

Besides fixing the actual problem this PR also moved the 'check for
timetrame availability' functionality to a dedicated private method to
be used by _add and _update.

Closes: LIB-253

jtojnar commented on hamster_lib/backends/sqlalchemy/storage.py:
```diff
@@ -926,6 +926,38 @@ def get_all(self):
+        """
+        Determine if a timeframe given by the passed fact is already occupied.
+
+        This method takes also such facts into account that start before and end
```
This should probably be re-worded – the first sentence does not have a natural word order and “exceeds” in the second one is not very clear. Something like the following might be better

> This method also takes into account such facts that start before and end after the fact in question. In that regard, it is a superset of what ``_get_all`` would return.

jtojnar commented on hamster_lib/backends/sqlalchemy/storage.py:
```diff
@@ -926,6 +926,38 @@ def get_all(self):
+
+        Args:
+            fact (Fact): The fact to check. Please note that the fact is expected to
+                have a ``start`` and ``end``.
```
→ an `end`

jtojnar commented on setup.cfg:
```diff
@@ -18,11 +18,11 @@ precision = 2
 [isort]
 not_skip = __init__.py
-known_third_party = appdirs, backports, faker,factory, faker, freezegun, future, hamster_lib, icalendar, past, pytest, pytest_factoryboy,
+known_third_party = appdirs, configparser, faker,factory, faker, freezegun, future, hamster_lib, icalendar, past, pytest, pytest_factoryboy,
```
`configparser` should probably be treated as standard library, since it is part of it on Python 3.

elbenfreund replied:
well.. its a nother py2 vs3 case :( under py2 it is a 3rd party lib, under python3 it is not
so there is no "right way" to do this i guess :/

jtojnar replied:
For future proofing, I would ignore Python 2 here.

elbenfreund replied:
I agree. To keep this hotfix branch clean and focused this has been place as a seperate [issue](https://projecthamster.atlassian.net/browse/LIB-256).

jtojnar commented on hamster_lib/backends/sqlalchemy/storage.py:
```diff
@@ -1095,10 +1122,16 @@ def _get_all(self, start=None, end=None, search_term='', partial=False):
+
+        Note:
+            This method will *NOT* return facts that start before and end after
+            (e.g. that span more than) the specified timeframe.
```
“id est” should be probably used here instead of “exemplī grātiā”.

jtojnar commented on hamster_lib/storage.py:
```diff
@@ -726,6 +726,8 @@ def get_all(self, start=None, end=None, filter_term=''):
             * ``search_term`` should be prefixable with ``not`` in order to invert matching.
             * This does only return proper facts and does not include any existing 'ongoing fact'.
+            * This method will *NOT* return facts that start before and end after
+              (e.g. that span more than) the specified timeframe.
```
ditto

jtojnar commented:
What is the relationship with #240?

jtojnar reviewed (commented):
Never mind, I thought there were some differences but it is just forward-port.

elbenfreund committed:
Fix import order

elbenfreund committed:
Merge branch 'bugfix/time_parse_seconds' into release/0.13.2

Previous versions of ``hamster-lib`` tried to be overly accomodating by
omitting seconds when parsing/displaying a times. This caused various
bugs and inconsistencies and is fixed by this commit.

This PR does two things:

- Fact representations now always shows time infomations as ``%H:%M:%S`` in
order to be more precise.
- rawfact / timeinfo parsing now accepts either ``%H:%M:%S`` or ``%H:%M`` as
valid time formats.

Closes: LIB-254

elbenfreund committed:
Bump version: 0.13.1 → 0.13.2

elbenfreund committed:
Add infos on fixes to changelog

jtojnar commented on HISTORY.rst:
```diff
@@ -3,6 +3,14 @@
+0.13.0 (2017-06-07)
+--------------------
+- Fix a bug that would not allways return all partial overlaps for
+  ``actManager._get_all``. This in effect fixes broken "timeframe availability"
```
Missing letter

jtojnar commented on HISTORY.rst:
```diff
@@ -3,6 +3,14 @@
 
+0.13.0 (2017-06-07)
+--------------------
+- Fix a bug that would not allways return all partial overlaps for
```
always

jtojnar commented on HISTORY.rst:
```diff
@@ -3,6 +3,14 @@
+  ``actManager._get_all``. This in effect fixes broken "timeframe availability"
+  tests. [LIB-253]
+- Remove obsolete references to ``backports.ConfigParser``. [LIB-255]
+- Properly deal with raw fact that do/do-not include seconds. [LIB-254]
```
fact to facts

elbenfreund committed:
Explicitly pin 'factory-boy' version

So far, ``factory-boy`` was not part of our explicit requirements
because we relied on ``pytest-factoryboy`` to pull it as on of its
dependencies. This however prooved error prone as recent ``factory-boy``
versions raise failures (at least in our context).
In order to avoid this we move from installing ``factory-boy``
implicitly to making it part of our requirements as well as pin its last
known functioning version.

For details please also the [relevant ``pytest-factoryboy``
issue](https://github.com/pytest-dev/pytest-factoryboy/issues/47)
